Mathlib/CategoryTheory/WithTerminal.lean

```diff
@@ -0,0 +1,485 @@
+/-
+Copyright (c) 2021 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+
+! This file was ported from Lean 3 source module category_theory.with_terminal
+! leanprover-community/mathlib commit 14b69e9f3c16630440a2cbd46f1ddad0d561dee7
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.CategoryTheory.Limits.Shapes.Terminal
+
+/-!
+
+# `with_initial` and `with_terminal`
+
+Given a category `C`, this file constructs two objects:
+1. `WithTerminal C`, the category built from `C` by formally adjoining a terminal object.
+2. `WithInitial C`, the category built from `C` by formally adjoining an initial object.
+
+The terminal resp. initial object is `WithTerminal.star` resp. `WithInitial.star`, and
+the proofs that these are terminal resp. initial are in `WithTerminal.star_terminal`
+and `WithInitial.star_initial`.
+
+The inclusion from `C` intro `WithTerminal C` resp. `WithInitial C` is denoted
+`WithTerminal.incl` resp. `WithInitial.incl`.
+
+The relevant constructions needed for the universal properties of these constructions are:
+1. `lift`, which lifts `F : C ⥤ D` to a functor from `WithTerminal C` resp. `WithInitial C` in
+  the case where an object `Z : D` is provided satisfying some additional conditions.
+2. `inclLift` shows that the composition of `lift` with `incl` is isomorphic to the
+  functor which was lifted.
+3. `liftUnique` provides the uniqueness property of `lift`.
+
+In addition to this, we provide `WithTerminal.map` and `WithInitinal.map` providing the
+functoriality of these constructions with respect to functors on the base categories.
+
+-/
+
+
+namespace CategoryTheory
+
+universe v u
+
+variable (C : Type u) [Category.{v} C]
+
+/-- Formally adjoin a terminal object to a category. -/
+inductive WithTerminal : Type u
+  | of : C → WithTerminal
+  | star : WithTerminal
+  deriving Inhabited
+#align category_theory.with_terminal CategoryTheory.WithTerminal
+
+/-- Formally adjoin an initial object to a category. -/
+inductive WithInitial : Type u
+  | of : C → WithInitial
+  | star : WithInitial
+  deriving Inhabited
+#align category_theory.with_initial CategoryTheory.WithInitial
+
+namespace WithTerminal
+
+-- porting note: no such tactic
+-- attribute [local tidy] tactic.case_bash
```

Comment:
```
attribute [local aesop safe cases] WithTerminal
```
does the job perfectly, it seems!

Reply:
You might prefer to add this to the `CategoryTheory` ruleset? But I think it is probably better for this attribute to be local, as it hopefully isn't needed after the API is set up.

Reply:
`aesop + cases` can now do some goals involving case analysis, but it still struggles to deal with `PEmpty.elim`